src/items/galleon.rs

```diff
@@ -0,0 +1,318 @@
+use macroquad::{
+    color,
+    experimental::{
+        animation::{AnimatedSprite, Animation},
+        collections::storage,
+        coroutines::{start_coroutine, Coroutine},
+        scene::{Handle, HandleUntyped, Node, RefMut},
+    },
+    prelude::*,
+    rand::gen_range,
+};
+
+use crate::{
+    capabilities,
+    components::{GunlikeAnimation, PhysicsBody, ThrowableItem},
+    nodes::Player,
+    Resources,
+};
+
+pub struct FlyingGalleon {
+    pub sprite: Texture2D,
+    pub pos: Vec2,
+    pub speed: Vec2,
+    pub facing: bool,
+    pub owner_id: u8,
+}
+
+impl FlyingGalleon {
+    pub const SPEED: f32 = 200.;
+    pub const WIDTH: f32 = 425.;
+    pub const HEIGHT: f32 = 390.;
+
+    pub fn new(owner_id: u8) -> FlyingGalleon {
+        let resources = storage::get::<Resources>();
+        let sprite = resources.items_textures["galleon/flying_galleon"];
+
+        let (pos, facing) = Self::start_position();
+        let dir = if facing {
+            Vec2::new(1., 0.)
+        } else {
+            Vec2::new(-1., 0.)
+        };
+
+        FlyingGalleon {
+            sprite,
+            pos,
+            speed: dir * Self::SPEED,
+            facing,
+            owner_id,
+        }
+    }
+
+    pub fn start_position() -> (Vec2, bool) {
+        let resources = storage::get::<Resources>();
+        let map_width =
+            resources.tiled_map.raw_tiled_map.tilewidth * resources.tiled_map.raw_tiled_map.width;
+        let map_height =
+            resources.tiled_map.raw_tiled_map.tileheight * resources.tiled_map.raw_tiled_map.height;
+
+        let facing = gen_range(0, 2) == 0;
+        let pos = Vec2::new(
+            if facing {
+                -Self::WIDTH
+            } else {
+                map_width as f32
+            },
+            gen_range(0., map_height as f32 - Self::HEIGHT),
+        );
+
+        (pos, facing)
+    }
+
+    pub fn update(&mut self) -> bool {
+        self.pos += self.speed * get_frame_time();
+
+        {
+            let resources = storage::get::<Resources>();
+            let map_width = (resources.tiled_map.raw_tiled_map.tilewidth
+                * resources.tiled_map.raw_tiled_map.width) as f32;
+            if self.pos.x + Self::WIDTH < -map_width * 0.5 || self.pos.x > map_width * 1.5 {
```

Comment:
What's the reason for not using the following condition?

```rs
self.pos.x + Self::WIDTH <= 0 || self.pos.x >= map_width
```

if there is a specific reason, I think it'd be good to add it to a comment, since it's not obvious when reading the code.

Reply:
When both players are on the far left or far right of the map, the outside is visible, if I changed to the code above then it would be visible when the Flying Galleon vanished, that's why I put it to vanish after it crossed half the map width off bounds.

Reply:
Right :+1: